[CDAP-18024] Add Support for Running as Non-Root User in Kubernetes

dli357 opened the pull request:
This feature leverages Kubernetes's [SecurityContext](https://kubernetes.io/docs/tasks/configure-pod-container/security-context/) feature in order to set a user ID, group ID, and filesystem access group ID for all the pods managed by CDAP operator.

The user ID and group ID define what user/group the pod will run with, and the filesystem access group ID defines which groups the volumes mounted onto the pod are accessible for. The filesystem access group ID is automatically applied to the user running in the pod.

For CDAP to function properly, the user assigned to ID 1000 must also be defined a valid user ID and group ID using the `useradd` and `groupadd` commands.

This is dependent on the following CDAP PR: https://github.com/cdapio/cdap/pull/13432

wyzhang commented on controllers/constants.go:
```diff
@@ -49,7 +49,7 @@ const (
 	confExploreEnabled        = "explore.enabled"
 	confLocalDataDirKey       = "local.data.dir"
-	confLocalDataDirVal       = "/data"
+	confLocalDataDirVal       = "/home/cdap/data"
```
Adding a comment saying that operator assumes that a user "cdap" exists in the image for this to worker properly.

dli357 replied:
Fixed by removing the hardcoded value.

wyzhang commented on controllers/constants.go:
```diff
@@ -64,6 +64,11 @@ const (
+	// kubernetes security context
+	defaultSecurityContextUID   = 1000
+	defaultSecurityContextGID   = 1000
+	defaultSecurityContextFSGID = 2000
```
As discussed offline, would be better to make security settings configurable via cdap-cr. When they are unset, default to current behavior of running as root.

dli357 replied:
Fixed by making it configurable.

wyzhang commented on controllers/cdapmaster_controller.go:
```diff
@@ -122,8 +122,6 @@ func ApplyDefaults(resource interface{}) {
 		spec.Config = make(map[string]string)
 	}
-	// Set the local data directory
-	spec.Config[confLocalDataDirKey] = confLocalDataDirVal
```
set local data dir if absent in the cdap-cr's config section

dli357 replied:
Done

wyzhang commented on templates/cdap-deployment.yaml:
```diff
@@ -110,6 +117,12 @@ spec:
             {{end}}
+          {{if $.Base.SecurityContext}}
+          securityContext:
+            privileged: {{$.Base.SecurityContext.Privileged}}
```
If some fields of security are left unspecified, I think we will use default value for them. Just wanted to make sure we thought about it. Default values look fine to me. Please double check

dli357 replied:
Good point, we need to be careful of the default behaviors. I've gone ahead and modified it so that:
 - If the integer values (runAsUser, runAsGroup, and fsGroup) are not specified (or zero) in the cdap-cr, no value is specified in the resulting yaml.
 - If the boolean values (allowPrivilegeEscalation, privileged, runAsNonRoot, readOnlyFilesystem) are not specified in the cdap-cr, they are explicitly set to default values defined by the Kubernetes API, i.e. allowPrivilegeEscalation = true, privileged = false, runAsNonRoot = false, readOnlyFilesystem = false.

dli357 committed:
[CDAP-18024] Add support for running as non-root user
[CDAP-18024] Allow security context and local data directory to be configurable in cdap-operator
